### PR TITLE
MSEARCH-4 Implement filters and facets

### PR DESCRIPTION
**Pull request by pfilippovEPAM**
- Added support in metadata for facet/filter fields
- Extracted from CqlSearchQueryConverter redundant staff
- Implemented method that flatten tree-like elasticsearch requests
- Implemented filtering using existing CQL
- Implemented facet operation without filters
- Implemented facet operation by specified fields with filters
- Implemented facet by tags.tagList using alias instanceTags
- Added integration tests for filter and facet operations

### CODE REVIEW

**Review by bohdan-suprun — APPROVED** [review]

File: src/main/resources/swagger.api/mod-search.yaml [hunk 1]
```diff
@@ -37,6 +28,26 @@ paths:
...
+      description: Provides list of facets for instance
+      parameters:
+        - $ref: '#/components/parameters/cql-facet-request'
+        - $ref: '#/components/parameters/x-okapi-tenant-header'
+      responses:
+        '200':
+          description: 'Facet search result'
+          content:
+            application/json:
+              example: examples/facetResult.sample
```
Comment (hjiebsco) [inline src/main/resources/swagger.api/mod-search.yaml:43]: Where is `facetResult.sample`?

File: src/main/resources/swagger.api/mod-search.yaml [hunk 2]
```diff
@@ -217,31 +225,20 @@ components:
...
-        maximum: 2147483647
-        type: integer
-    search-pageable-limit:
-      name: limit
+        $ref: '#/components/schemas/cqlSearchRequest'
+    cql-facet-request:
+      name: cqlFacetRequest
       in: query
-      description: Limit the number of elements returned in the response
+      description: A CQL query string with search conditions.
```
Comment (hjiebsco) [inline src/main/resources/swagger.api/mod-search.yaml:238]: Update description to mention `a list of facet names`

**Comment by hjiebsco** [comment]
Feel free to merge to master, but please add the missing `facetResult.sample` first.

File: src/main/java/org/folio/search/repository/SearchRepository.java [hunk 3]
```diff
@@ -28,41 +21,23 @@
...
   private final RestHighLevelClient elasticsearchClient;
-  private final ElasticsearchHitConverter elasticsearchHitConverter;
 
   /**
    * Executes request to elasticsearch and returns search result with related documents.
    *
-   * @param queryBuilder elasticsearch query as {@link QueryBuilder} object.
-   * @param cqlSearchRequest search request as {@link CqlSearchRequest} object.
+   * @param searchSource elasticsearch search source as {@link SearchSourceBuilder} object.
+   * @param resourceRequest resource request as {@link ResourceRequest} object.
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/java/org/folio/search/repository/SearchRepository.java]: Reordered.

File: src/main/java/org/folio/search/service/FacetService.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  private static void cleanUpFacetSearchSource(SearchSourceBuilder searchSource) {
+    var query = searchSource.query();
+    if (isBoolQuery(query)) {
+      ((BoolQueryBuilder) query).filter().clear();
+    }
+    if (CollectionUtils.isNotEmpty(searchSource.sorts())) {
+      var sorts = searchSource.sorts();
+      sorts.clear();
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/java/org/folio/search/service/FacetService.java]: Combined, thanks.

File: src/main/resources/swagger.api/schemas/request/cqlSearchRequest.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      "minimum": 0,
+      "default": 0
+    },
+    "expandAll": {
+      "type": "boolean",
+      "description": "Whether to return only basic properties or entire instance.",
+      "default": false
+    }
+  },
+  "required": [ "query", "expandAll" ]
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/resources/swagger.api/schemas/request/cqlSearchRequest.json]: It is not required, removed from the required list.

File: src/main/resources/swagger.api/schemas/response/searchResult.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "Facet result response",
```
[the remark this answers is not in the document]
Comment (bohdan-suprun) [reply src/main/resources/swagger.api/schemas/response/searchResult.json]: Renamed to `Search result response`.

File: src/main/resources/swagger.api/mod-search.yaml (hunk 1 above)
Comment (bohdan-suprun) [reply src/main/resources/swagger.api/mod-search.yaml:43]: It is added now.

File: src/main/resources/swagger.api/mod-search.yaml (hunk 2 above)
Comment (bohdan-suprun) [reply src/main/resources/swagger.api/mod-search.yaml]: Renamed to `A CQL query string with search conditions and facet names.`

**Pull request merged**